File: src/tree/binary-tree/binary-search-tree/binarySearchTree.cs
using System;
using System.Collections.Generic;

namespace BinarySearchTree
{
    public class BinarySearchTree
    {
        public enum Mode { preorder, inorder, postorder, levelorder };

        internal class Node
        {
            internal int _data;
            internal Node? _parent;
            internal Node? _left;
            internal Node? _right;

            internal Node(in int d)
            {
                _data = d;
                _parent = _left = _right = null;
            }
        }

        private Node? _root;

        public BinarySearchTree()
        {
            _root = null;
        }

        // Insert
        public bool Insert(in int d)
        {
            if (_root == null)
            {
                _root = new Node(d);
                return true;
            }

            // Find place and insert.
            Node cur = _root;
            while (true)
            {
                if (d < cur._data)
                {
                    if (cur._left == null)
                    {
                        cur._left = new Node(d);
                        cur._left._parent = cur;

                        return true;
                    }
                    else cur = cur._left;
                }
                else if (cur._data < d)
                {
                    if (cur._right == null)
                    {
                        cur._right = new Node(d);
                        cur._right._parent = cur;

                        return true;
                    }
                    else cur = cur._right;
                }
                else return false;
            }
        }

        // Remove elements from tree
        private Node? _findMin(Node? t)
        {
            Node? cur = t;
            while (cur != null && cur._left != null)
                cur = cur._left;

            return cur;
        }

        private bool isRoot(in Node? n)
        {
            return n._parent == null ? true : false;
        }

        public bool Detree(in int d)
        {
            if (_root == null) return false;

            Node cur = _root;

            while (cur != null)
            {
                if (d < cur._data)
                    cur = cur._left;
                else if (cur._data < d)
                    cur = cur._right;
                else break;
            }
            if (cur == null) return false;

            if (cur._left != null && cur._right != null) // two children
            {
                Node temp = _findMin(cur._right);
                int value = temp._data;

                Detree(value);
                cur._data = value;
            }
            else if (cur._left == null && cur._right == null) // No child
            {
                if (!isRoot(cur))
                {
                    if (cur == cur._parent._left) cur._parent._left = null;
                    else cur._parent._right = null;
                }
                else _root = null;
            }
            else // One child;
            {
                Node? whichChild = cur._left == null ? cur._right : cur._left;

                if (!isRoot(cur))
                {
                    if (cur._parent._left == cur) cur._parent._left = whichChild;
                    else cur._parent._right = whichChild;
                }
                else _root = whichChild;
            }

            return true;
        }

        private void PreorderTraverse(Node? t)
        {
            if (t == null) return;
            Console.Write($"{t._data} ");
            PreorderTraverse(t._left);
            PreorderTraverse(t._right);
        }

        private void InorderTraverse(Node? t)
        {
            if (t == null) return;
            InorderTraverse(t._left);
            Console.Write($"{t._data} ");
            InorderTraverse(t._right);
        }

        private void PostorderTraverse(Node? t)
        {
            if (t == null) return;
            PostorderTraverse(t._left);
            PostorderTraverse(t._right);
            Console.Write($"{t._data} ");
        }

        private void LevelorderTraverse(Node? t)
        {
            Queue<Node?> q = new Queue<Node?>();

            q.Enqueue(t);
            while (q.Count > 0)
            {
                Node? n = q.Dequeue();
                Console.Write($"{n._data} ");

                if (n._left != null) q.Enqueue(n._left);
                if (n._right != null) q.Enqueue(n._right);
            }
        }

        public void Traverse(Mode m)
        {
            switch (m)
            {
                case Mode.preorder:
                    PreorderTraverse(_root);
                    break;

                case Mode.inorder:
                    InorderTraverse(_root);
                    break;

                case Mode.postorder:
                    PostorderTraverse(_root);
                    break;

                case Mode.levelorder:
                    LevelorderTraverse(_root);
                    break;
            }
            Console.WriteLine();
        }
    }

    public static class MainApp
    {
        public static void Main()
        {
            var tree = new BinarySearchTree();

            tree.Insert(491);
            tree.Insert(198);
            tree.Insert(566);
            tree.Insert(814);
            tree.Insert(802);
            tree.Insert(423);
            tree.Insert(826);
            tree.Insert(555);
            tree.Insert(818);
            tree.Insert(802);

            tree.Traverse(BinarySearchTree.Mode.preorder);

            tree.Detree(555);
            tree.Traverse(BinarySearchTree.Mode.preorder);
            tree.Detree(814);
            Console.WriteLine();
            Console.Write("Preorder: ");
            tree.Traverse(BinarySearchTree.Mode.preorder);
            Console.Write("Inorder: ");
            tree.Traverse(BinarySearchTree.Mode.inorder);
            Console.Write("Postorder: ");
            tree.Traverse(BinarySearchTree.Mode.postorder);
            Console.Write("Levelorder: ");
            tree.Traverse(BinarySearchTree.Mode.levelorder);
        }
    }
}

File: src/tree/binary-tree/binary-search-tree/binarySearchTree.c
// 이진 탐색 트리 Binary Search Tree without external nodes

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "binarySearchTree.h"

node* creat_node(int data)
{
    node* rt = (node*)malloc(sizeof(node));
    rt->parent = rt->left = rt->right = NULL;
    rt->data = data;
    
    return rt;
}

void insert(node** root, node* parent, int data)
{
    if (!(*root))
    {
        (*root) = creat_node(data);
        (*root)->parent = parent;
        return;
    }

    if (data < (*root)->data) insert(&((*root)->left), *root, data);
    else if (data == (*root)->data) return;
    else insert(&((*root)->right), *root, data);
}

node* find_successor(node* n)
{
    while (n->left)
        n = n->left;
    return n;
}

int detree(node** root, node* parent, int target)
{
    if (!(*root)) return -1;

    if (target < (*root)->data)
        return detree(&((*root)->left), *root, target);
    else if ((*root)->data == target)
    {
        node* cur = (*root)->left;
        int data = (*root)->data;
        if (cur) cur = (*root)->right;

        if (!cur) // The node has only one child or no child
        {
            node* successor = (cur == (*root)->left) ? (*root)->right : (*root)->left;
            if (successor) successor->parent = (*root)->parent;
            free(*root);
            *root = successor;
        }
        else // The node has two children
        {
            node* successor = find_successor((*root)->right);
            (*root)->data = successor->data;
            
            if (successor->parent->left == successor)
                successor->parent->left = successor->right;
            else successor->parent->right = successor->right;
            if (successor->right) successor->right->parent = successor->parent;
            free(successor);
        }

        return data;
    }
    else return detree(&((*root)->right), *root, target);
}

void traversal(node* root, void (*method)(node*)) // function pointer
{
    method(root);
    putchar('\n');
}

void inorder(node* root)
{
    if (!root) return;

    if (root->left) inorder(root->left);
    printf("%d ", root->data);
    if (root->right) inorder(root->right);
}

void preorder(node* root)
{
    if (!root) return;

    printf("%d ", root->data);
    if (root->left) preorder(root->left);
    if (root->right) preorder(root->right);
}

void postorder(node* root)
{
    if (!root) return;

    if (root->left) postorder(root->left);
    if (root->right) postorder(root->right);
    printf("%d ", root->data);
}

void free_all(node* root)
{
    if (!root) return;

    if (root->left) free_all(root->left);
    if (root->right) free_all(root->right);
    free(root);
}

int main(void)
{
    node* root = NULL;

    insert(&root, NULL, 123);
    insert(&root, NULL, 87);
    insert(&root, NULL, 124);
    insert(&root, NULL, 523);
    traversal(root, inorder);

    detree(&root, NULL, 123);
    traversal(root, inorder);
    detree(&root, NULL, 2222);

    insert(&root, NULL, 89);
    insert(&root, NULL, 88);
    insert(&root, NULL, 91);
    traversal(root, inorder);
    detree(&root, NULL, 89);
    traversal(root, inorder);
    traversal(root, preorder);
    traversal(root, postorder);

    free_all(root);
    return 0;
}